src/util/delayed_trigger.js

```diff
@@ -0,0 +1,82 @@
+/* Copyright (C) 2023 NooBaa */
+'use strict';
+
+const Semaphore = require('./semaphore');
+
+/**
+ * DelayedTrigger is used to invoke a batch handling from smaller triggers.
+ * For example collecting stats into a variable and sending it all once in a while.
+ */
+class DelayedTrigger {
+
+    /**
+     * @param {{
+     *      delay: number;
+     *      max_retries?: number;
+     *      on_trigger: () => Promise<any>;
+     * }} params
+     */
+    constructor({ delay, on_trigger, max_retries = Infinity }) {
+        this._delay = delay;
+        this._on_trigger = on_trigger;
+        this._max_retries = max_retries;
+        this._num_retries = 0;
+        this._timer = null;
+        this._sem = new Semaphore(1);
+    }
+
+    trigger() {
+        // when the timer is set we already have a trigger in place
+        if (this._timer) return;
```

Comment:
yes, when the timer is scheduled, for example if the delay is 10 seconds, then during those 10 seconds we will hit this first condition and bail, because the timer is already set. After 10 seconds the timer will be called and will unset this field and immediately go and acquire the semaphore. So unsetting the timer allows new stats updates to register a new timer.